### ISSUE TITLE
SDK - Components - typing.List and typing.Dict are not supported by func_to_container_op and behave differently in python 3.6 and python 3.7

### ISSUE THREAD

**Issue by devashishshankar** [kubeflow/pipelines#4144]
### What steps did you take:

In python3.6, the json serialization for List/ Dict types works well. But not so in python3.7. I built the following pipeline using both a python3.6 venv and a python3.7 venv (Both venv had only kfp package installed):

```
from typing import List, Dict

from kfp import dsl
from kfp.compiler import Compiler
from kfp.components import func_to_container_op


def print_op(a: List[str], b: Dict):
    print(a, type(a))
    print(b, type(b))


print_op_factory = func_to_container_op(print_op)


@dsl.pipeline(name="Test Ser", description="Test ser")
def my_pipeline(a: List[str], b: Dict):
    print_op_ = print_op_factory(a, b)


Compiler().compile(my_pipeline, "./test-ser-pipeline.tar.gz")

```

### What happened:

This is the diff of generated pipeline.yaml:

diff pipeline-36.yaml pipeline-37.yaml

```
<   annotations: {pipelines.kubeflow.org/kfp_sdk_version: 0.5.1, pipelines.kubeflow.org/pipeline_compilation_time: '2020-07-04T17:38:25.436101',
---
>   annotations: {pipelines.kubeflow.org/kfp_sdk_version: 0.5.1, pipelines.kubeflow.org/pipeline_compilation_time: '2020-07-04T17:40:39.523850',
7c7,8
<       "a", "type": "List"}, {"name": "b", "type": "Dict"}], "name": "Test Ser"}'}
---
>       "a", "type": "typing.List[str]"}, {"name": "b", "type": "typing.Dict"}], "name":
>       "Test Ser"}'}
24d24
<         import json
27,28c27,28
<         _parser.add_argument("--a", dest="a", type=json.loads, required=True, default=argparse.SUPPRESS)
<         _parser.add_argument("--b", dest="b", type=json.loads, required=True, default=argparse.SUPPRESS)
---
>         _parser.add_argument("--a", dest="a", type=str, required=True, default=argparse.SUPPRESS)
>         _parser.add_argument("--b", dest="b", type=str, required=True, default=argparse.SUPPRESS)
54c54,55
<           "type": "List"}, {"name": "b", "type": "Dict"}], "name": "Print op"}'}
---
>           "type": "typing.List[str]"}, {"name": "b", "type": "typing.Dict"}], "name":
>           "Print op"}'}
```

I believe python3.6 generated pipeline is the correct one - as per https://github.com/kubeflow/pipelines/pull/1945/files. In my main pipeline, I faced a serialization issue when i generated it using python3.7

I believe there were some changes in the typing API in py3.7 - which could be causing this - but not sure.

### What did you expect to happen:

(Explained above)

### Environment:

(Explained above) - my env was basically a new virtualenv with just kfp installed.

How did you deploy Kubeflow Pipelines (KFP)?
CLI

KFP version: kfctl_gcp_iap.v1.0.2.yaml
KFP SDK version:
```
kfp                           0.5.1
kfp-server-api                0.5.0
```

### Anything else you would like to add:

/kind bug
/area sdk

**Comment by Ark-kun** [comment]
Thank you for the report @devashishshankar .
It's an interesting issue. There is currently no support for types from the `typing` module. Only `list` and `dict` are supported. And by "supported" I mean that they're converted to the canonical `"JsonArray"` and `"JsonObject"` type names and serializers/deserializers are used.
In python 3.6, `typing.List` is a `type` and has `.name == 'List'`. But in Python 3.7 `typing.List` is not a `type`.

Would you be fine using `list` and `dict for now`?

`typing.*` members are a bit hard to work with (they're converted to `GenericMeta` or `GenericAlias` and not easy to restore) and they have many differences between 3.5, 3.6 and 3.7.

**Comment by devashishshankar** [comment]
> Would you be fine using list and dict for now?

Yeah, that works for now. I've fixed the version to 3.6 for now.

I know this is a painful issue. I faced this some time back - and I ended up using [typing-inspect](https://github.com/ilevkivskyi/typing_inspect) library. It provides a uniform API for python 3.5+. But even after using this - I know it isn't very trivial. There were still lots of if conditions in my code to check various possibilities :)
